### PR TITLE
Use Shedlock for account reminder emails

### LINKED ISSUE #3382: Verify the use of threading in ReminderService
### Description
the following lines have been flagged for bad practices

https://github.com/CDCgov/prime-simplereport/blob/2efa52c9f30483f8b20d8412208a1a8c211ba4b3/backend/src/main/java/gov/cdc/usds/simplereport/service/ReminderService.java#L87

https://github.com/CDCgov/prime-simplereport/blob/2efa52c9f30483f8b20d8412208a1a8c211ba4b3/backend/src/main/java/gov/cdc/usds/simplereport/service/ReminderService.java#L84

### Proposed solution 
Verify the correct use of threading in ReminderService

### Additional context
from fortify scan, contact @emmastephenson  @zdeveloper for more context

### PR DESCRIPTION
**Pull request by mehansen**
# BACKEND PULL REQUEST

## Related Issue

Resolves #3382 

## Changes Proposed

Thread management and specifically using `sleep()` is bad practice in web apps and error-prone. The code doing this was originally added because multiple instances of our API would fire off the account reminder emails for unverified orgs, causing duplicate emails. The original solution was to sleep while holding an advisory lock so that other instances would not have a chance to grab it even if there was any kind of clock difference (the method can complete very quickly if there are no emails to send). Instead we are using Shedlock, which uses a table in our db to store info such that a method can be locked for a specified amount of time after it's called.

* set up and configure ShedLock
* switch from using `ThreadPoolTaskScheduler` to using the `@Scheduled` annotation for the account reminder emails
* use ShedLock to limit method execution to 1x per day and remove code that uses the advisory lock

## Additional Information

* https://www.baeldung.com/spring-scheduled-tasks
* https://www.baeldung.com/shedlock-spring
* https://github.com/lukas-krecan/ShedLock#duration-specification
* https://github.com/CDCgov/prime-simplereport/pull/2641
* https://github.com/CDCgov/prime-simplereport/pull/2768

following up with a PR for adding a low-urgency alert if the method is executed more or less than 1x a day to catch any issues with misconfigurations since this isn't covered by our tests

## Testing

~~please double-check my Liquibase changes!~~

this [test](https://github.com/CDCgov/prime-simplereport/blob/fe13710d29380f991c44fdc132747c4163735fe6/backend/src/test/java/gov/cdc/usds/simplereport/service/ReminderServiceTest.java#L60) verifies that we still prevent duplicate executions 

to test in a lower, I set the schedule to run every 2 min and set the ShedLock config to hold the lock for 3 min minimum and verified that the method ran every 4 min (skipping times when the method lock was still held by the previous run) 

## Checklist for Primary Reviewer
- [x] Any large-scale changes have been deployed to `test`, `dev`, or `pentest` and smoke tested
- [x] Changes comply with the SimpleReport Style Guide
- [x] Any dependencies introduced have been vetted and discussed

### CODE REVIEW

File: backend/src/main/java/gov/cdc/usds/simplereport/db/repository/AdvisoryLockManager.java [hunk 1]
```diff
@@ -16,22 +15,6 @@ public interface AdvisoryLockManager {
...
-   * @param lock the specific lock we are trying to obtain (usually a constant in the implementing
-   *     repository or the service that calls it).
-   */
-  @Query(
-      nativeQuery = true,
-      // can't tell hibernate that Types.OTHER is a "void" result in this case:
-      // just cast it to text
-      value = "select cast(pg_advisory_xact_lock(:lockCategory, :lock) as text)")
-  void waitForTransactionLock(int lockCategory, int lock);
-
```
Comment (mehansen) [inline backend/src/main/java/gov/cdc/usds/simplereport/db/repository/AdvisoryLockManager.java:34]: this was already not being used - not related to my changes but I cleaned it up anyway

**Comment by mehansen** [comment]
moving this back to a draft while I figure out what's causing the `OutOfMemory` errors in our backend tests. seems to be something related to Bean instantiation so I may have misconfigured something with ShedLock

File: backend/src/main/resources/db/changelog/db.changelog-master.yaml [hunk 2]
```diff
@@ -4131,4 +4131,39 @@ databaseChangeLog:
                   descending: true
       rollback:
         - dropIndex:
-            indexName: ix__patient_link__created_at
\ No newline at end of file
+            indexName: ix__patient_link__created_at
+  - changeSet:
+      id: create-shedlock-table
```
Comment (emmastephenson) [inline backend/src/main/resources/db/changelog/db.changelog-master.yaml:4136]: I know this PR is still in draft, but just a note that it's best practice for database changes to go into a separate PR - it makes rollbacks a bit easier/safer than bundling the changes together.

Comment (mehansen) [reply backend/src/main/resources/db/changelog/db.changelog-master.yaml:4184]: created #4989 for this - thanks for the call out!

**Review by emmastephenson — APPROVED** [review]
LGTM, this looks great! A much more easily-readable solution than the ScheduledTasksService too.

File: backend/src/main/java/gov/cdc/usds/simplereport/service/ReminderService.java [hunk 3]
```diff
@@ -32,19 +33,25 @@ public ReminderService(OrganizationQueueRepository orgQueueRepo, EmailService em
...
   }
 
+  /**
+   * Wrapper method for sending account reminder emails so automation can call the inner method
+   * without hitting the lock or conditions.
+   */
+  @Scheduled(cron = "0 0 1 * * *", zone = "America/New_York")
+  @SchedulerLock(
+      name = "ReminderService_sendAccountReminderEmails",
+      lockAtLeastFor = "PT30S",
```
Comment (zdeveloper) [inline backend/src/main/java/gov/cdc/usds/simplereport/service/ReminderService.java:43]: two questions
what is the PT in front of the time? the [docs](https://github.com/lukas-krecan/ShedLock#:~:text=%40SchedulerLock(name%20%3D%20%22scheduledTaskName%22%2C%20lockAtMostFor%20%3D%20%2214m%22%2C%20lockAtLeastFor%20%3D%20%2214m%22)) just have numbers

**Review by zdeveloper — COMMENTED** [review]
LGTM! would love to see it on an azure env before final approval

**Comment by mehansen** [comment]
@zdeveloper deployed on dev4! the scheduled task did not run last night because `simple-report.id-verification-reminders.enabled` is set to false for that env. I did do the following in a lower before putting up the PR: 
> to test in a lower, I set the schedule to run every 2 min and set the ShedLock config to hold the lock for 3 min minimum and verified that the method ran every 4 min (skipping times when the method lock was still held by the previous run)

here's the commit where I did that: https://github.com/CDCgov/prime-simplereport/pull/4891/commits/31f23119df52c7e2d236662a8787ef412aa4e4f1

is there anything else you'd like to see tested on a lower? it might be fun to deploy somewhere I have metabase access and see the table populated but the above wouldn't have worked without populating the table so not very worried about it

**Review by zdeveloper — APPROVED** [review]
Thank you!
